### PR TITLE
Added TypeScript type definitions

### PR DESCRIPTION
**Pull request by ivanz**
Hi,

This pull request adds type definitions (for TypeScript and editors that support type definitions) and integrates them in package.json. I have added a couple of tests as well to check a couple of things when using the typings.

---

This change is [Reviewable](https://reviewable.io/reviews/astorije/chai-immutable/63)

### CODE REVIEW

**Comment by astorije** [comment]
This is huge, thanks! I'll need some time reviewing this as I have never been nearby TypeScript nor any of the tooling you used here.

I can't promise when, but I'll review this carefully.

File: chai-immutable.d.ts [hunk 1]
```diff
@@ -0,0 +1,3 @@
+import * as chai from './typing/lib/chai-immutable';
+
+export = chai;
```
Comment (astorije) [inline chai-immutable.d.ts]: Missing EOF here :-)

File: tslint.json [hunk 2]
```diff
@@ -0,0 +1,7 @@
+{
+  "extends": "tslint-config-typings",
+  "rules": {
+    "no-trailing-whitespace": false,
+    "eofline": false
```
Comment (astorije) [inline tslint.json]: Is this related to previous comment?

**Comment by ivanz** [comment]
Review status: 0 of 14 files reviewed at latest revision, 2 unresolved discussions.

---

*[chai-immutable.d.ts, line 3 \[r1\]](https://reviewable.io:443/reviews/astorije/chai-immutable/63#-KJu-q-y2ajI8W8Nz0jA-r1-3:-KJu-z3eaaqmSV73xqk3:1758359711) ([raw file](https://github.com/astorije/chai-immutable/blob/4d3527e10bf7a0128075440678fb403d39a04ae0/chai-immutable.d.ts#L3)):*
<details><summary><i>Previously, astorije (Jérémie Astori) wrote…</i></summary>
> Missing EOF here :-)
</details>
Fixed.

---

*[tslint.json, line 5 \[r1\]](https://reviewable.io:443/reviews/astorije/chai-immutable/63#-KJu-q-y2ajI8W8Nz0jC-r1-5:-KJu0CqLDlBNVs_pSRz0:-785425264) ([raw file](https://github.com/astorije/chai-immutable/blob/4d3527e10bf7a0128075440678fb403d39a04ae0/tslint.json#L5)):*
<details><summary><i>Previously, astorije (Jérémie Astori) wrote…</i></summary>
> Is this related to comment above?
</details>
Yeah - I've changed this so that the whitespace and eof rules are enabled.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/astorije/chai-immutable/63)*

**Comment by astorije** [comment]
> Fixed.

You sure? I still see the missing EOF in the files tab.

**Comment by ivanz** [comment]
I think GitHub is not displaying them? I can see them locally and tslint passes. If I remove a line it fails. To be honest the new lines don't matter for the typings, because they are just metadata for development-time only (during TypeScript compilation) and it's not something that gets bundled or used during run-time.

File: .gitignore [hunk 3]
```diff
@@ -1,3 +1,10 @@
+.vscode
```
Comment (astorije) [inline .gitignore]: Can you remove this line from the PR? It's a system-specific ignored file and should be in people's global gitconfig (this also why there is no `.DS_Store` in this file even though I'm running on macOS at the moment).

File: README.md [hunk 4]
```diff
@@ -311,3 +311,10 @@ Asserts that the immutable collection has the expected size.
 assert.sizeOf(List.of(1, 2, 3), 3);
 assert.sizeOf(new List(), 0);
 ```
+
+
+## TypeScript type definitions
+
+The `chai-immutable` module contains TypeScript type definitions, which should automatically be picked up by TypeScript.
```
Comment (astorije) [inline README.md]: s/The `chai-immutable` module/This plugin/

File: README.md [hunk 5]
```diff
@@ -311,3 +311,10 @@ Asserts that the immutable collection has the expected size.
 assert.sizeOf(List.of(1, 2, 3), 3);
 assert.sizeOf(new List(), 0);
 ```
+
+
+## TypeScript type definitions
```
Comment (astorije) [inline README.md]: Could you move this section right above the API reference please?

File: .editorconfig [hunk 6]
```diff
@@ -0,0 +1,5 @@
+root = true
+
+[*.{js,json}]
+indent_style = space
+indent_size = 2
```
Comment (astorije) [inline .editorconfig]: Never used this, wouldn't this file be a simple translation of the lint config file(s)? (I'll move from JSCS to ESLint at some point)

File: chai-immutable.d.ts (hunk 1 above)
Comment (astorije) [inline chai-immutable.d.ts]: On top of the missing EOF, would it make sense to call it `chaiImmutable` or `chaiImmutableTyping` instead of `chai`? (I am still totally inexperienced with TS so just trying to figure out what's in this var 😅)

File: package.json [hunk 7]
```diff
@@ -30,6 +35,12 @@
     "name": "Jérémie Astori",
     "url": "https://jeremie.astori.fr"
   },
+  "contributors": [
+    {
+      "name": "Ivan Zlatev",
+      "url": "http://ivanz.com"
+    }
+  ],
```
Comment (astorije) [inline package.json]: I'm not 100% sure about this, but I do appreciate the help from contributors a lot!
For those who are wondering, I'll add other contributor names once this PR is merged, and if I happen to ever receive hundreds of contributions, I'll rethink this 😄.

File: package.json [hunk 8]
```diff
@@ -44,6 +55,10 @@
     "immutable": "^3.7.5",
     "istanbul": "^0.4.3",
     "jscs": "^2.5.0",
-    "mocha": "^2.4.5"
+    "mocha": "^2.4.5",
+    "typescript": "^1.8.10",
+    "typings": "^1.0.4",
+    "tslint": "*",
+    "tslint-config-typings": "^0.2.0"
```
Comment (astorije) [inline package.json]: That's quite a lot of dependencies added here. Could you add a quick comment about why you use them all for?

Is `typescript` actually necessary in the devDependencies? I'd be annoyed if I had to install an extra 4 dependencies (~40M) to use this small plugin on a JS package.

While we're here, can you bump versions of these dependencies and make sure it still works?

File: typing/lib/chai-immutable.d.ts [hunk 9]
```diff
@@ -0,0 +1,11 @@
+import {Utils} from 'chai/lib/Utils';
+import {} from './Assertion';
+
+// chai-immutable exports a single function (is a callable? module).
```
Comment (astorije) [inline typing/lib/chai-immutable.d.ts]: Comment needs fixing?

File: typing/test/Assertation.specs.ts [hunk 10]
```diff
@@ -0,0 +1,30 @@
+import {expect, use} from 'chai';
```
Comment (astorije) [inline typing/test/Assertation.specs.ts]: Why are the tests in 2 different files instead of just `chai-immutable.specs.ts`?

File: package.json [hunk 11]
```diff
@@ -6,13 +6,18 @@
...
     "test": "npm run test-mocha; npm run test-phantomjs; npm run check-style",
     "test-mocha": "mocha",
     "test-phantomjs": "echo 'mocha-phantomjs is temporarily disabled.' # mocha-phantomjs test/index.html",
     "coverage": "istanbul cover _mocha",
-    "coveralls": "npm run coverage && cat ./coverage/lcov.info | coveralls"
+    "coveralls": "npm run coverage && cat ./coverage/lcov.info | coveralls",
+    "build-typings": "tsc -p .",
+    "test-typings": "npm run build-typings && npm run check-style-typings && mocha typing/build/*.specs.js",
+    "check-style-typings": "tslint typing/lib/**/*.ts typing/test/**/*.ts",
+    "prepublish": "typings install"
```
Comment (astorije) [inline package.json]: This will run at each npm publish (i.e. every time I push a tag to GitHub that gets picked up by Travis), is that on purpose?

File: typings.json [hunk 12]
```diff
@@ -0,0 +1,9 @@
+{
+  "dependencies": {
+    "chai": "registry:npm/chai#3.5.0+20160415060238",
+    "immutable": "registry:npm/immutable#3.7.6+20160411060006"
+  },
+  "globalDevDependencies": {
+    "mocha": "registry:env/mocha#2.2.5+20160321223601"
+  }
+}
```
Comment (astorije) [inline typings.json]: This is a bit cryptic to me, what does that mean when I bump dependency versions?
Also, there seems to be a discrepancy between versions here and in `package.json`.

In general, anything in particular I need to know when I add features to the API, publish a version, bump a dep, ...?

**Comment by ivanz** [comment]
Hey - I forgot about this pull request :) I am going to close this as it's no longer needed. The new version of Typescript introduced a nice way to pull and manage typings outside of JS libraries and  I saw that @jsebrech has already created some here: https://github.com/jsebrech/chai-immutable-typings

**Comment by astorije** [comment]
Argh, I wish I knew that before spending time reviewing :D
Alright, I know nothing about TypeScript, does that mean this plugin doesn't need any change/addition to be TypeScript-friendly?

Does it make sense to add a section in the README, and/or an example?

If I understand properly, you should deprecate https://github.com/ivanz/chai-immutable-typings and redirect to this plugin + https://github.com/jsebrech/chai-immutable-typings or the repo behind https://github.com/DefinitelyTyped/DefinitelyTyped/pull/13067 once it is merged.

**Comment by ivanz** [comment]
No changes needed to this repo unless you feel like you want to collaborate with @jsebench to pull his typing file into the repo and publish with the NPM package. Up to you guys.

**Comment by astorije** [comment]
Ping @jsebrech, what do you think is the right thing to do here?

**Comment by jsebrech** [comment]
You need definition files to pull a non-typescript library into a typescript codebase, but it doesn't matter whether those are included with the base library or distributed separately via definitelytyped / npm. So, this doesn't need to be rolled into chai-immutable, but it could be if someone felt up to maintaining it.

**Comment by astorije** [comment]
I don't know the first thing about TypeScript (at least for now) so it might make more sense to keep it in a repo that's managed by TypeScript folks.
However, feel free to open a PR that updates README to document where and how to use these typing files.
